awx/main/models/unified_jobs.py

```diff
@@ -357,18 +366,48 @@ def create_unified_job(self, **kwargs):
         if hasattr(self, 'survey_spec') and getattr(self, 'survey_enabled', False):
             password_list = self.survey_password_variables()
             hide_password_dict = getattr(unified_job, 'survey_passwords', {})
+            hide_password_dict.update(original_passwords)
             for password in password_list:
                 hide_password_dict[password] = REPLACE_STR
             unified_job.survey_passwords = hide_password_dict
 
         unified_job.save()
 
-        # Labels and extra credentials copied here
+        # Labels and credentials copied here
+        if kwargs.get('credentials'):
+            Credential = UnifiedJob._meta.get_field('credentials').related_model
+            cred_dict = Credential.unique_dict(self.credentials.all())
+            prompted_dict = Credential.unique_dict(kwargs['credentials'])
+            # combine prompted credentials with JT
+            cred_dict.update(prompted_dict)
+            kwargs['credentials'] = [cred for cred in cred_dict.values()]
+
         from awx.main.signals import disable_activity_stream
         with disable_activity_stream():
             copy_m2m_relationships(self, unified_job, fields, kwargs=kwargs)
+
+        if 'extra_vars' in kwargs:
+            unified_job.handle_extra_data(kwargs['extra_vars'])
+
+        if not getattr(self, '_deprecated_credential_launch', False):
+            # Create record of provided prompts for relaunch and rescheduling
+            unified_job.create_config_from_prompts(kwargs)
+
         return unified_job
 
+    @cached_subclassproperty
+    def ask_mapping(cls):
+        mapping = {}
+        for field in cls._meta.fields:
+            if not isinstance(field, AskForField):
+                continue
+            if field.allows_field == '__default__':
```

Comment:
Instead of using a special string here `__default__` and at https://github.com/ansible/awx/pull/740/files#diff-ea6da88b8c0cd3fbf7858ae87933b296R770, why not just use `None`?  That way the `default=...` argument is how you specify behavior that's *different* from the default.

Instead of building this mapping every time you need it and dealing with caching, why not construct it and store in the `AskForField` e.g.,

Change this code:
https://github.com/ansible/awx/pull/740/files#diff-ea6da88b8c0cd3fbf7858ae87933b296R770

...to do this:

```python
def __init__(self, allows_field=None, **kwargs):
    super(AskForField, self).__init__(**kwargs)
    self.allows_field = allows_field or self.name[len('ask_'):-len('_on_launch')]
```

Reply:
Probably need to add a property on the AskForField field class to complement that, like

```python
@property
def allows_field(self):
    return self._allows_field if self._allows_field is not None else self.name
```

Agree - something with more of this structure would look great.

We can't reference `self.name` on initialization, but it won't be a problem with this approach.

Reply:
Well...

```python
... else self.name[len('ask_'):-len('_on_launch')
```

Yes?

Reply:
ugh, yes, that. I only claim to meta-code in github comments.